### PR TITLE
In-memory Engine: integrate hybrid engine with TiKV

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
Integrate hybrid engine with TiKV. User can choose to use hybrid engine by set `memory_engine_enabled` in TiKV config.
```
This PR refactors some parts mainly in component/server to replace RocksEngine by generic type implementing KvEngine.
So we can make `HybridEngine` as a specific type that implements KvEngine be used by TiKV server, RaftKv, raftstore and so on.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Connor1996 — COMMENTED** [review]
Do we really need engine trait to use it? Maybe regard it as a cache, and just add a field in apply and region snapshot?

**Comment by SpadeA-Tang** [comment]
> Do we really need engine trait to use it? Maybe regard it as a cache, and just add a field in apply and region snapshot?

Use trait can make fewer changes to raftstore.

File: src/config/mod.rs [hunk 1]
```diff
@@ -3499,6 +3501,7 @@ impl Default for TikvConfig {
             abort_on_panic: false,
             memory_usage_limit: None,
             memory_usage_high_water: 0.9,
+            hybrid_engine_enabled: false,
```
Comment (tonyxuqqi) [inline src/config/mod.rs:3504]: The name is too hard for user to understand. I think we may just use the region-cache-memory-quota to enable or disable the feature. By default, it's empty which means it's disabled.

Comment (SpadeA-Tang) [reply src/config/mod.rs:3504]: Good suggestion.

File: components/snap_recovery/src/services.rs [hunk 2]
```diff
@@ -136,7 +142,7 @@ impl<ER: RaftEngine> RecoveryService<ER> {
         // config rocksdb l0 to optimize the restore
         // also for massive data applied during the restore, it easy to reach the write
         // stop
-        let db = engines.kv.clone();
+        let db: &RocksEngine = engines.kv.get_disk_engine();
```
Comment (tonyxuqqi) [inline components/snap_recovery/src/services.rs:145]: I think it's better to use kv directly and internally it passes the call to disk_engine. 
The principle is that keep minimal code change outside of the engine components.

**Review by tonyxuqqi — APPROVED** [review]

**Comment by SpadeA-Tang** [comment]
PTAL thanks @Connor1996

File: components/server/src/server.rs [hunk 3]
```diff
@@ -209,9 +220,33 @@ pub fn run_tikv(
 
     dispatch_api_version!(config.storage.api_version(), {
         if !config.raft_engine.enable {
-            run_impl::<RocksEngine, API>(config, service_event_tx, service_event_rx)
+            if config.region_cache_memory_limit == ReadableSize(0) {
+                run_impl::<RocksEngine, RocksEngine, API>(
```
Comment (Connor1996) [inline components/server/src/server.rs:224]: I suggest always using HybridEngine and check the region cache limit innerly

Comment (SpadeA-Tang) [reply components/server/src/server.rs:224]: After discussed offline, we think this is riskless once it is not enabled.

Comment (Connor1996) [reply components/server/src/server.rs:224]: But using `HybridEngine` and when memory engine is not enabled, the code is actually a wrapper on RocksEngine without doing anything else

Comment (SpadeA-Tang) [reply components/server/src/server.rs:224]: It still needs to do some logics to check whether the memory engine can be used or something.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**